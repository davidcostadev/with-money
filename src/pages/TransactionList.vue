<template>
  <div>
    <toolbar v-on:update:stateMonth="getTransactions" />
    <md-tabs>
      <md-tab md-label="Fixas"></md-tab>
      <md-tab md-label="VariÃ¡veis"></md-tab>
    </md-tabs>
    <md-table>
      <md-table-row>
        <md-table-head>Nome</md-table-head>
        <md-table-head>Valor</md-table-head>
        <md-table-head>Pago?</md-table-head>
        <md-table-head>Data</md-table-head>
      </md-table-row>
      <md-table-row
        v-for="transaction in transactions"
        :key="transaction.id"
        @click="onSelectTransaction(transaction)"
      >
        <md-table-cell>{{transaction.name}}</md-table-cell>
        <md-table-cell>{{transaction.value}}</md-table-cell>
        <md-table-cell>{{transaction.transationDate}}</md-table-cell>
        <md-table-cell>{{transaction.isPaid ? 'Paid' : 'Not Paid'}}</md-table-cell>
      </md-table-row>
    </md-table>
  </div>
</template>

<script>
import Moment from 'moment';
import { mapGetters } from 'vuex';
import Transaction from '../services/transactions';
import Toolbar from '../components/Toolbar';

export default {
  components: { Toolbar },
  data() {
    return {
      transactions: [],
    };
  },
  computed: {
    ...mapGetters(['state_month']),
  },
  methods: {
    onSelectTransaction({ id }) {
      this.$router.push(`/transaction/${id}`);
    },
    async getTransactions() {
      const start = Moment(this.state_month).startOf('month').toISOString();
      const end = Moment(this.state_month).endOf('month').toISOString();
      const { data } = await Transaction.getTransactions({
        createdAt: [
          start,
          end,
        ].join(','),
      });

      this.transactions = data;
    },
  },
  created() {
    this.getTransactions();
  },
};
</script>

<style>
  .md-content {
    padding: 16px;
  }
</style>
