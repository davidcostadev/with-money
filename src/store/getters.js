export default {
  user(state) {
    return state.user;
  },
  current_month(state) {
    return state.current_month;
  },
  state_month(state) {
    return state.state_month;
  },
  state_month_str(state) {
    return state.state_month_str;
  },
  next_month_str(state) {
    return state.next_month_str;
  },
  previous_month_str(state) {
    return state.previous_month_str;
  },
};
