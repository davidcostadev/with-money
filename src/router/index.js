import Vue from 'vue';
import Router from 'vue-router';
import Login from '@/pages/Login';
import TransactionEdit from '@/pages/TransactionEdit';
import TransactionList from '@/pages/TransactionList';
import store from '../store';

Vue.use(Router);

const insecure = (to, from, next) => {
  const token = window.localStorage.getItem('token');

  if (!store.getters.user && !token) {
    next();
    return;
  }

  next('/');
};

const secure = (to, from, next) => {
  const token = window.localStorage.getItem('token');

  if (store.getters.user || token) {
    next();
    return;
  }

  next('/login');
};

export default new Router({
  routes: [
    {
      path: '/login',
      name: 'login',
      component: Login,
      beforeEnter: insecure,
    },
    {
      path: '/',
      name: 'TransactionList',
      component: TransactionList,
      beforeEnter: secure,
    },
    {
      path: '/transaction/:id',
      name: 'TransactionEdit',
      component: TransactionEdit,
      beforeEnter: secure,
    },
  ],
});
