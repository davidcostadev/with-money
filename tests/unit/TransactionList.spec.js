import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import VueMaterial from 'vue-material';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import TransactionList from '../../src/pages/TransactionList';
// import store from '../../src/store';
// import Vuex from 'vuex';
import Moment from 'moment';
import router from '../../src/router';
import getters from '../../src/store/getters';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueMaterial);
localVue.use(VueRouter);

const Category = {
  name: 'Lanche',
};

const transactions = [
  {
    id: 3,
    AccountId: 1,
    name: 'AlmoçoAlmoço 2',
    value: '15.00',
    type: 'out',
    isPaid: false,
    transactionDate: '2018-07-02',
    createdAt: '2018-08-04T21:26:56.000Z',
    updatedAt: '2018-08-04T21:27:07.000Z',
    Category,
  },
  {
    id: 2,
    AccountId: 2,
    name: 'Coxinha',
    value: '3.13',
    type: 'out',
    isPaid: false,
    transactionDate: '2018-08-04',
    createdAt: '2018-08-04T19:08:54.000Z',
    updatedAt: '2018-08-04T19:08:54.000Z',
    Category,
  },
  {
    id: 1,
    AccountId: 1,
    name: 'Lanche na Sonia',
    value: '30.90',
    type: 'out',
    isPaid: true,
    transactionDate: '2018-08-04',
    createdAt: '2018-08-04T19:08:54.000Z',
    updatedAt: '2018-08-04T19:08:54.000Z',
    Category,
  },
];

const mock = {
  data: transactions,
  pagination: {
    totalItems: 3,
    currentPage: 1,
    perPage: 100,
    totalPages: 1,
    nextPage: null,
    previousPage: null,
  },
};

jest.mock('../../src/services/transactions', () => ({
  getTransactions: () => Promise.resolve(mock),
}));

describe('TransactionList Component', () => {
  let now;
  let store;

  beforeAll(() => {
    global.localStorage = {
      getItem: jest.fn().mockReturnValue(JSON.stringify({
        id: '1',
        name: 'name',
      })),
    };
    now = Date.now;
    Date.now = () => 1532473578215;

    store = new Vuex.Store({
      state: {
        current_month: Moment(),
        state_month: Moment(),
        state_month_str: Moment().format('MMMM YY'),
        next_month_str: Moment().add(1, 'month').format('MMMM YY'),
        previous_month_str: Moment().subtract(1, 'month').format('MMMM YY'),
      },
      getters,
    });
  });

  afterAll(() => {
    Date.now = now;
  });

  describe('Render', () => {
    it('default', () => {
      const wrapper = shallowMount(TransactionList, { store, localVue, router });

      expect(wrapper.isVueInstance).toBeTruthy();
      expect(wrapper.element).toMatchSnapshot();
    });

    it('with transactions', () => {
      const stubs = {
        MdTab: '<div class="md-tab-stubbed" />',
      };

      const wrapper = mount(TransactionList, { stubs, store, localVue, router });
      wrapper.setData({
        transactions,
      });

      expect(wrapper.isVueInstance).toBeTruthy();
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
