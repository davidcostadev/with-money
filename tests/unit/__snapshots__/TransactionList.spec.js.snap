// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList Component Render default 1`] = `
<div
  class="full-height"
>
  <toolbar-stub />
   
  <div
    class="content-scroll"
  >
    <mdtabs-stub
      id="tabs-type"
      mdactivetab="tab-out"
      mdalignment="left"
      mdelevation="0"
    >
      <mdtab-stub
        id="tab-in"
        mdlabel="Entradas"
        mdtemplatedata="[object Object]"
      />
       
      <mdtab-stub
        id="tab-out"
        mdlabel="Saidas"
        mdtemplatedata="[object Object]"
      />
    </mdtabs-stub>
     
    <mdtable-stub
      md-card=""
      mdcard=""
    >
      <mdtablerow-stub>
        <mdtablehead-stub>
          Data
        </mdtablehead-stub>
         
        <mdtablehead-stub>
          Nome
        </mdtablehead-stub>
         
        <mdtablehead-stub>
          Valor
        </mdtablehead-stub>
         
        <mdtablehead-stub>
          Categoria
        </mdtablehead-stub>
         
        <mdtablehead-stub
          style="width: 75px;"
        >
          Pago?
        </mdtablehead-stub>
      </mdtablerow-stub>
       
    </mdtable-stub>
     
    <mdcard-stub
      class="calcs"
    >
      <mdcardcontent-stub>
        <div
          class="md-layout"
        >
          <div
            class="md-layout-item"
          >
            Total Entradas
          </div>
           
          <div
            class="md-layout-item text-green"
          >
            R$ 0,00
          </div>
        </div>
         
        <div
          class="md-layout"
        >
          <div
            class="md-layout-item"
          >
            Total Saídas
          </div>
           
          <div
            class="md-layout-item text-red"
          >
            R$ 0,00
          </div>
        </div>
         
        <div
          class="md-layout"
        >
          <div
            class="md-layout-item"
          >
            Total
          </div>
           
          <div
            class="md-layout-item"
          >
            R$ 0,00
          </div>
        </div>
      </mdcardcontent-stub>
    </mdcard-stub>
     
    <mdcard-stub
      class="calcs"
    >
      <mdcardcontent-stub>
        <progressbartransaction-stub
          amount="NaN"
          partial="0"
          title="Entradas"
          total="0"
          type="in"
        />
         
        <progressbartransaction-stub
          amount="NaN"
          partial="0"
          title="Saídas"
          total="0"
          type="out"
        />
      </mdcardcontent-stub>
    </mdcard-stub>
  </div>
   
  <mdbutton-stub
    class="md-fab md-primary menu-speed"
    to="[object Object]"
    type="button"
  >
    <mdicon-stub>
      note_add
    </mdicon-stub>
  </mdbutton-stub>
</div>
`;

exports[`TransactionList Component Render with transactions 1`] = `
<div
  class="full-height"
>
  <div
    class="md-toolbar md-theme-default md-elevation-4"
  >
    <button
      class="md-button md-theme-default"
      id="btn-previus-month"
      type="button"
    >
      <div
        class="md-ripple"
      >
        <div
          class="md-button-content"
        >
          <i
            class="md-icon md-icon-font md-theme-default"
          >
            chevron_left
          </i>
           June 18
  
        </div>
         
      </div>
    </button>
     
    <h3
      class="md-title"
      style="text-align: center;"
    >
      July 18
    </h3>
     
    <button
      class="md-button md-theme-default"
      id="btn-next-month"
      type="button"
    >
      <div
        class="md-ripple"
      >
        <div
          class="md-button-content"
        >
          <i
            class="md-icon md-icon-font md-theme-default"
          >
            chevron_right
          </i>
           August 18
  
        </div>
         
      </div>
    </button>
  </div>
   
  <div
    class="content-scroll"
  >
    <div
      class="md-tabs md-alignment-left md-no-transition md-theme-default"
      id="tabs-type"
    >
      <div
        class="md-tabs-navigation md-elevation-0"
      >
         
        <span
          class="md-tabs-indicator"
        />
      </div>
       
      <div
        class="md-content md-tabs-content md-theme-default"
        style="height: 0px; display: none;"
      >
        <div
          class="md-tabs-container"
        >
          <div
            class="md-tab-stubbed"
            id="tab-in"
            md-label="Entradas"
          />
           
          <div
            class="md-tab-stubbed"
            id="tab-out"
            md-label="Saidas"
          />
        </div>
      </div>
    </div>
     
    <div
      class="md-card md-table md-theme-default"
      md-card=""
    >
       
      <!---->
       
      <!---->
       
      <div
        class="md-content md-table-content md-scrollbar md-theme-default"
      >
        <table>
          <!---->
           
          <tbody>
            <tr
              class="md-table-row"
            >
              <!---->
               
              <th
                class="md-table-head"
              >
                <div
                  class="md-table-head-container"
                >
                  <div
                    class="md-table-head-label"
                  >
                    Data
                  </div>
                </div>
              </th>
               
              <th
                class="md-table-head"
              >
                <div
                  class="md-table-head-container"
                >
                  <div
                    class="md-table-head-label"
                  >
                    Nome
                  </div>
                </div>
              </th>
               
              <th
                class="md-table-head"
              >
                <div
                  class="md-table-head-container"
                >
                  <div
                    class="md-table-head-label"
                  >
                    Valor
                  </div>
                </div>
              </th>
               
              <th
                class="md-table-head"
              >
                <div
                  class="md-table-head-container"
                >
                  <div
                    class="md-table-head-label"
                  >
                    Categoria
                  </div>
                </div>
              </th>
               
              <th
                class="md-table-head"
                style="width: 75px;"
              >
                <div
                  class="md-table-head-container"
                >
                  <div
                    class="md-table-head-label"
                  >
                    Pago?
                  </div>
                </div>
              </th>
            </tr>
             
            <tr
              class="md-table-row"
              id="transaction-3"
            >
              <!---->
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  07/02
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  AlmoçoAlmoço 2
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  15.00
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  Lanche
                </div>
              </td>
               
              <td
                class="md-table-cell md-table-cell-center"
              >
                <div
                  class="md-table-cell-container"
                >
                  <i
                    class="md-icon md-icon-font md-theme-default"
                  >
                    lens
                  </i>
                </div>
              </td>
            </tr>
            <tr
              class="md-table-row"
              id="transaction-2"
            >
              <!---->
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  08/04
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  Coxinha
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  3.13
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  Lanche
                </div>
              </td>
               
              <td
                class="md-table-cell md-table-cell-center"
              >
                <div
                  class="md-table-cell-container"
                >
                  <i
                    class="md-icon md-icon-font md-theme-default"
                  >
                    lens
                  </i>
                </div>
              </td>
            </tr>
            <tr
              class="md-table-row"
              id="transaction-1"
            >
              <!---->
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  08/04
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  Lanche na Sonia
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  30.90
                </div>
              </td>
               
              <td
                class="md-table-cell"
              >
                <div
                  class="md-table-cell-container"
                >
                  Lanche
                </div>
              </td>
               
              <td
                class="md-table-cell md-table-cell-center"
              >
                <div
                  class="md-table-cell-container"
                >
                  <i
                    class="md-icon md-icon-font md-theme-default"
                  >
                    check_circle
                  </i>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
         
      </div>
       
      <!---->
    </div>
     
    <div
      class="md-card calcs md-theme-default"
    >
      <div
        class="md-card-content"
      >
        <div
          class="md-layout"
        >
          <div
            class="md-layout-item"
          >
            Total Entradas
          </div>
           
          <div
            class="md-layout-item text-green"
          >
            R$ 0,00
          </div>
        </div>
         
        <div
          class="md-layout"
        >
          <div
            class="md-layout-item"
          >
            Total Saídas
          </div>
           
          <div
            class="md-layout-item text-red"
          >
            R$ 49,03
          </div>
        </div>
         
        <div
          class="md-layout"
        >
          <div
            class="md-layout-item"
          >
            Total
          </div>
           
          <div
            class="md-layout-item"
          >
            -R$ 49,03
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="md-card calcs md-theme-default"
    >
      <div
        class="md-card-content"
      >
        <div
          class="progress-bar-transaction"
        >
          <div
            class="progress-bar-transaction-title"
          >
            Entradas
          </div>
           
          <div
            appear=""
            class="md-progress-bar md-determinate md-theme-default"
            name="md-progress-bar"
          >
            <div
              class="md-progress-bar-track"
              style="width: 0%;"
            />
             
            <div
              class="md-progress-bar-fill"
            />
             
            <div
              class="md-progress-bar-buffer"
              style="left: calc(0% + 8px)"
            />
          </div>
           
          <div
            class="md-layout"
          >
            <div
              class="md-layout-item"
            >
              <span>
                Realizado: 
              </span>
               
              <span
                class="text-in"
              >
                R$ 0,00
              </span>
            </div>
             
            <div
              class="md-layout-item"
            >
              <span>
                Previsto: 
              </span>
               
              <span
                class="text-partial"
              >
                R$ 0,00
              </span>
            </div>
          </div>
        </div>
         
        <div
          class="progress-bar-transaction"
        >
          <div
            class="progress-bar-transaction-title"
          >
            Saídas
          </div>
           
          <div
            appear=""
            class="md-progress-bar md-determinate md-theme-default md-accent"
            name="md-progress-bar"
          >
            <div
              class="md-progress-bar-track"
              style="width: 0%;"
            />
             
            <div
              class="md-progress-bar-fill"
              style="width: 63.02263920048949%;"
            />
             
            <div
              class="md-progress-bar-buffer"
              style="left: calc(0% + 8px)"
            />
          </div>
           
          <div
            class="md-layout"
          >
            <div
              class="md-layout-item"
            >
              <span>
                Realizado: 
              </span>
               
              <span
                class="text-out"
              >
                R$ 30,90
              </span>
            </div>
             
            <div
              class="md-layout-item"
            >
              <span>
                Previsto: 
              </span>
               
              <span
                class="text-partial"
              >
                R$ 49,03
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <a
    class="md-button md-fab md-primary menu-speed md-theme-default"
    href="#/transaction-new?type=out"
  >
    <div
      class="md-ripple"
    >
      <div
        class="md-button-content"
      >
        <i
          class="md-icon md-icon-font md-theme-default"
        >
          note_add
        </i>
      </div>
       
    </div>
  </a>
</div>
`;
